### PR TITLE
Improve rate limiter

### PR DESCRIPTION
**Pull request by atxr**

### CODE REVIEW

**Comment by ctmbl** [comment]
This PR's parent is #19 which out-scoped its issue
from https://github.com/iScsc/RootPythia/pull/19#issuecomment-1610217982

TODO:
- [ ] Tests (for 429, 4XX, timeout and max_retry) @ctmbl 
- [ ] Send discord messages (@ctmbl I don't really know how to handle this so if you can take a look, it would be with pleasure)
- [ ] Format and pylint

**Comment by ctmbl** [comment]
@atxr unit tests didn't pass :confused: the Action has been killed by GH it was probably hanging in some await that never retrieves, could you take a look?

**Comment by atxr** [comment]
I'll have a look!

**Comment by ctmbl** [comment]
@atxr I solved it with [22e832b](https://github.com/iScsc/RootPythia/pull/26/commits/22e832bb5176496ae12ae9925bf268a08f8b95de) [8ba538f](https://github.com/iScsc/RootPythia/pull/26/commits/8ba538fcd0dc669e04a83528a4e9e9000c22080d) and [1e94141](https://github.com/iScsc/RootPythia/pull/26/commits/1e94141936102527aa83caacf6c26569b31bd689)

**Comment by ctmbl** [comment]
should fix #24 too

**Review by atxr — COMMENTED** [review]
Very happy about our work on this PR!
🚀

File: src/api/rate_limiter.py [hunk 1]
```diff
@@ -27,122 +49,168 @@ class RateLimiter:
...
+                try:
+                    time_to_wait = int(resp.headers["Retry-After"])
+                except (KeyError, ValueError) as exc:
+                    raise RateLimiterError(
+                        request,
+                        self.logger.error,
+                        "Too many requests (429) and cannot parse headers",
+                    ) from exc
+
+                raise RLErrorWithPause(
```
Comment (FireFrozen94) [inline src/api/rate_limiter.py:111]: time_to_wait is define l.103 in the try, so if there is an exception with the header reading, won't it crash when trying to execute the raise RLErrorWithPause with a time_to_wait outdated or uninitialized ?

Comment (ctmbl) [reply src/api/rate_limiter.py:111]: That's why we `except` this case:
if the parsing fail, we raise an exception **without** `time_to_wait` L105 (and it "exits" the fucntion)
if it succeeds we end up here and raise an exxcpetion **with** it

**Review by Xlitoni — APPROVED** [review]
Personally I have no remarks to make, this is nice :+1:

**Comment by atxr** [comment]
@ctmbl go merge?

**Pull request merged**